proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,330 @@
+  The API of ``ghc-internals`` is fully under the control of the GHC team, and of no direct interest to the CLC — only its effects on the API of base.
+
+- ``base``: as now, whose API is curated by CLC.
+  Depends on ``ghc-internals``, and hence on ``ghc-bignum`` and ``ghc-prim``.
+
+- ``ghc-experimental``, initially empty, depends on ``base`` and on ``ghc-internals``.
+  Functions and data types here are intended to have their ultimate home in base, but while they are settling down they are subject to much weaker stability guarantees.
+  Generally, new functions and types introduced in GHC Proposals would start their life here.
+  Example: new type families and type constructors for tuples, `GHC Proposal #475 <https://github.com/ghc-proposals/ghc-proposals/pull/475>`__.
+
+  Another example: future APIs to access RTS statistics, which are fairly stable and user-exposed, but which are (by design) coupled closely to GHC's runtime and hence may change.
+
+  As its name suggests, the API of ``ghc-experimental`` is curated by the GHC team, although the CLC is willing to offer (non-binding) opinions, if consulted.
+
+All three packages conform rigorously to the PVP.
+(But see Section 5.3)
+
+Some observations about this structure:
+
+- We should use all possible social and technical means to discourage people from depending directly on ``ghc-internals``, because if such dependencies become frequent and ossified, it will lead to future pain when the API changes.
+  Saying "we told you not to rely on it" may be true, but won't lessen that pain.
+  What mechanisms could we use?
+
+  - The name ``ghc-internals`` is a pretty strong signal all by itself.
+
+  - Cabal description and README explains how it is intended used (and not used).
+
+  - Hoogle could (by default anyway) never show stuff from ``ghc-internals``.
+
+  - Do not upload Haddocks for ``ghc-internals`` to Hackage.
+    (Ditto ``ghc-prim``.) Need to make sure that if someone wants to follow the Haddock source-code link to (say) Functor, they should still find it regardless of where it is actually defined.
+
+  - We could consider issuing a warning if you say ``-package ghc-internals`` (or ``ghc-bignum`` or ``ghc-prim``), one that was hard to silence.
+    Since we can have module-level ``WARNING`` pragmas with custom categories, one way to realise this would be to pick a category and add such pragmas to every module in the relevant packages, though we might want to do something more systematic.
+    The text of the warnings could encourage users to
+
+    - switch to a function exposed by base, and/or
+    - petition the CLC to expose this super-useful function from base.
+
+  - ``cabal check`` (a per-package check) could warn on packages that use ``ghc-internals``.
+
+  - ...what else?
```

Comment:
> We should use all possible social and technical means to discourage people from depending directly on ``ghc-internals``

Really all? Including GHC/Cabal simply refusing to allow packages called  `ghc-internals` if the package currently built isn't `base` or another internal whitelisted package?

But it says “discourage”, not “prevent”, which sounds reasonable.


Maybe more realistic: hackage could reject uploads of packages that depend on `ghc-internals`?

Reply:
> But it says “discourage”, not “prevent”, which sounds reasonable.

Yes that is key. If people want to tinker with the internals of some open source software, we should not stop them. But we should make sure they are very aware they are stepping outside the stable safe zone. And likewise we ought to someday make anyone that uses the tinker's software (transitively!) also aware they are exposed to instability.

Reply:
(The difference between `ghc-prim` and `ghc-internal` also isn't clear. But one answer is

> GHC devs can structure (and restructure) their internals however they like, including dividing definitions *arbitrarily* between `ghc-prim` and `ghc-internals` if they so desire, so this question doesn't need to be answered in this document.

Another answer is

> `ghc-prim` is (mostly, perhaps should be entirely) auto-generated docs for code that doesn't actually exist in textual form (it's just hard-coded in GHC), whereas `ghc-internals` is human-written code.

)